src/Sync/AsyncFileMutex.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace Amp\File\Sync;
+
+use Amp\Coroutine;
+use Amp\Delayed;
+use Amp\Promise;
+use Amp\Sync\Lock;
+use Amp\Sync\Mutex;
+use Amp\Sync\SyncException;
+use function Amp\File\open;
+use function Amp\File\unlink;
+
+final class AsyncFileMutex implements Mutex
+{
+    private const LATENCY_TIMEOUT = 10;
+
+    /** @var string The full path to the lock file. */
+    private $fileName;
+
+    /**
+     * @param string|null $fileName Name of temporary file to use as a mutex.
+     */
+    public function __construct(string $fileName)
+    {
+        $this->fileName = $fileName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function acquire(): Promise
+    {
+        return new Coroutine($this->doAcquire());
+    }
+
+    /**
+     * @coroutine
+     *
+     * @return \Generator
+     */
+    private function doAcquire(): \Generator
+    {
+        // Try to create the lock file. If the file already exists, someone else
+        // has the lock, so set an asynchronous timer and try again.
+        while (($file = yield open($this->fileName, 'x')) === false) {
+            yield new Delayed(self::LATENCY_TIMEOUT);
+        }
+
+        // Return a lock object that can be used to release the lock on the mutex.
+        $lock = new Lock(0, function (): void {
+            $this->release();
+        });
+
+        yield $file->close();
+
+        return $lock;
+    }
+
+    /**
+     * Releases the lock on the mutex.
+     *
+     * @throws SyncException If the unlock operation failed.
+     */
+    private function release(): void
+    {
+        unlink($this->fileName)->onResolve(
+            function (?\Throwable $exception): void {
+                if ($exception !== null) {
+                    throw new SyncException(
+                        'Failed to unlock the mutex file: ' . $this->file,
+                        0,
+                        $exception
+                    );
+                }
+            }
+        );
```

Comment:
This should use a coroutine. In the current code, the `SyncException` is thrown into the event loop and halts the entire application if no error handler is set.

Reply:
@kelunik I don't understand how to use a coroutine here... What happens instead if a coroutine is used? Will the exception be silently ignored?

Reply:
I forgot that `release` doesn't return a promise, so nevermind.

Reply:
Perhaps it should then? Breaking the entire event loop because a lock could not be released without making it catchable might not be desired.

Reply:
We'd need to release `amphp/sync` v2 for that. We could change it to a warning instead of an exception, but most long running applications should use an error handler, so they don't unexpectedly crash.